behaviour/graql/language/rule-validation.feature

```diff
@@ -23,6 +23,509 @@ Feature: Graql Rule Validation
+        $clyde isa person, has name "Clyde";
+      }, then {
+        (criminal: $bonnie, sidekick: $clyde) isa partners-in-crime;
+      };
+      """
+    Then the integrity is validated
+
+
+  Scenario: when a rule infers a relation with an incorrect roleplayer, an error is thrown
+    Then graql define throws
+      """
+      define
+      partners-in-crime sub relation, relates criminal, relates sidekick;
+      person plays criminal;
+      bonnie-and-clyde-are-partners-in-crime sub rule,
+      when {
+        $bonnie isa person, has name "Bonnie";
+        $clyde isa person, has name "Clyde";
+      }, then {
+        (criminal: $bonnie, sidekick: $clyde) isa partners-in-crime;
+      };
+      """
+    Then the integrity is validated
+
+
+  @ignore
+  # TODO: re-enable when rules cannot infer abstract relations
+  Scenario: when a rule infers an abstract relation, an error is thrown
+    Then graql define throws
+      """
+      define
+      partners-in-crime sub relation, abstract, relates criminal, relates sidekick;
+      person plays criminal, plays sidekick;
+      bonnie-and-clyde-are-partners-in-crime sub rule,
+      when {
+        $bonnie isa person, has name "Bonnie";
+        $clyde isa person, has name "Clyde";
+      }, then {
+        (criminal: $bonnie, sidekick: $clyde) isa partners-in-crime;
+      };
+      """
+    Then the integrity is validated
+
+
+  @ignore
+  # TODO: re-enable when rules cannot infer abstract attribute values
+  Scenario: when a rule infers an abstract attribute value, an error is thrown
+    Then graql define throws
+      """
+      define
+      number-of-devices sub attribute, value long, abstract;
+      person has number-of-devices;
+      karl-is-allergic-to-technology sub rule,
+      when {
+        $karl isa person, has name "Karl";
+      }, then {
+        $karl has number-of-devices 0;
+      };
+      """
+    Then the integrity is validated
+
+
+  Scenario: when a rule negates its conclusion in the `when`, causing a loop, an error is thrown
+    Then graql define throws
+      """
+      define
+      there-are-no-unemployed sub rule,
+      when {
+        $person isa person;
+        not {
+          (employee: $person) isa employment;
+        };
+      }, then {
+        (employee: $person) isa employment;
+      };
+      """
+    Then the integrity is validated
+
+
+  @ignore
+  # TODO: re-enable when subrules are not allowed
+  Scenario: attempting to `sub` another rule label throws an error
+    Then graql define throws
+    """
+    define
+    nickname sub name;
+    person has nickname;
+    robert-has-nickname-bob sub rule,
+    when {
+      $p isa person, has name "Robert";
+    }, then {
+      $p has nickname "Bob";
+    };
+    robert-has-nickname-bobby sub robert-has-nickname-bob,
+    when {
+      $p isa person, has name "Robert";
+    }, then {
+      $p has nickname "Bobby";
+    };
+    """
+    Then the integrity is validated
```

Comment:
loads of this stuff is not necessarily negation, so we should update the PR description to say this correspondingly

Reply:
Already included!